llama_index/llms/xinference.py

```diff
@@ -0,0 +1,153 @@
+            **base_kwargs,
+            **self._model_description,
+        }
+        return model_kwargs
+
+    def _get_input_dict(self, prompt: str, **kwargs: Any) -> Dict[str, Any]:
+        return {"prompt": prompt, **self._model_kwargs, **kwargs}
+
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        prompt = messages[-1].content if len(messages) > 0 else ""
+        history = [xinference_message_to_history(message) for message in messages[:-1]]
+        response_text = self._generator.chat(
+            prompt=prompt,
+            chat_history=history,
+            generate_config={"stream": False, "temperature": self.temperature},
+        )["choices"][0]["message"]["content"]
+        response = ChatResponse(
+            message=ChatMessage(
+                role=MessageRole.ASSISTANT,
+                content=response_text,
+            ),
+            delta=None,
+        )
+        return response
+
+    def stream_chat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponseGen:
+        prompt = messages[-1].content if len(messages) > 0 else ""
+        history = [xinference_message_to_history(message) for message in messages[:-1]]
+        response_iter = self._generator.chat(
+            prompt=prompt,
+            chat_history=history,
+            generate_config={"stream": True, "temperature": self.temperature},
+        )
+
+        def gen() -> ChatResponseGen:
+            text = ""
+            for c in response_iter:
+                delta = c["choices"][0]["delta"].get("content", "")
+                text += delta
+                yield ChatResponse(
+                    message=ChatMessage(
+                        role=MessageRole.ASSISTANT,
+                        content=text,
+                    ),
+                    delta=delta,
+                )
+
+        return gen()
+
+    def complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
+        response_text = self._generator.chat(
+            prompt=prompt,
+            chat_history=None,
+            generate_config={"stream": False, "temperature": self.temperature},
+        )["choices"][0]["message"]["content"]
+        response = CompletionResponse(
+            delta=None,
+            text=response_text,
+        )
+        return response
+
+    def stream_complete(self, prompt: str, **kwargs: Any) -> CompletionResponseGen:
+        response_iter = self._generator.chat(
+            prompt=prompt,
+            chat_history=None,
+            generate_config={"stream": True, "temperature": self.temperature},
+        )
+
+        def gen() -> CompletionResponseGen:
+            text = ""
+            for c in response_iter:
+                delta = c["choices"][0]["delta"].get("content", "")
+                text += delta
+                yield CompletionResponse(
+                    delta=delta,
+                    text=text,
+                )
+
+        return gen()
```

Comment:
Hmm, I wasn't actually able to get streaming working with a query engine. Did you?

Reply:
Would you mind providing a minimal code example? We were told to follow the structure of [replicate.py](https://github.com/jerryjliu/llama_index/blob/main/llama_index/llms/replicate.py) in the issue we started, so we tested `stream_complete` by replicating the code and temporarily reroute the `complete` function to `stream_complete` in the same way:

```
    def complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
        response_gen = self.stream_complete(prompt, **kwargs)
        response_list = list(response_gen)
        final_response = response_list[-1]
        final_response.delta = None
        return final_response
```

That worked when we tested it in the demo notebooks. Please let us know if you encountered any errors or if there are other ways to test the `stream_complete` function and we will address it ASAP.